charms/jupyter-ui/src/charm.py

```diff
@@ -25,139 +39,190 @@ def __init__(self, msg, status_type=None):
-                                    {
-                                        "path": "spawner_ui_config.yaml",
-                                        "content": Path("src/spawner_ui_config.yaml").read_text(),
-                                    }
-                                ],
-                            },
-                            {
-                                "name": "logos",
-                                "mountPath": "/src/apps/default/static/assets/logos",
-                                "files": [
-                                    {
-                                        "path": name,
-                                        "content": content,
-                                    }
-                                    for name, content in yaml.safe_load(
-                                        Path("src/logos-configmap.yaml").read_text()
-                                    )["data"].items()
-                                ],
-                            },
-                        ],
-                    }
-                ],
             },
+        }
+
+        return Layer(layer_config)
+
+    def _update_layer(self) -> None:
+        """Update the Pebble configuration layer (if changed)."""
+        current_layer = self.container.get_plan()
+        new_layer = self._jupyter_ui_layer
+        if current_layer.services != new_layer.services:
+            self.unit.status = MaintenanceStatus("Applying new pebble layer")
+            self.container.add_layer(self._container_name, new_layer, combine=True)
+            try:
+                self.logger.info("Pebble plan updated with new configuration, replaning")
+                self.container.replan()
+            except ChangeError:
+                raise ErrorWithStatus("Failed to replan", BlockedStatus)
+
+    def _upload_files_to_container(self):
+        """Upload required files to container."""
+        self.container.push(
+            "/etc/config/spawner_ui_config.yaml",
+            "spawner_ui_config.yaml",
+            make_dirs=True,
         )
-        self.model.unit.status = ActiveStatus()
+        for file_name, file_content in yaml.safe_load(
+            Path("src/logos-configmap.yaml").read_text()
+        )["data"].items():
+            logo_file = "/src/apps/default/static/assets/logos/" + file_name
+            self.container.push(
+                logo_file,
+                file_content,
+                make_dirs=True,
+            )
+
+    def _deploy_k8s_resources(self) -> None:
+        """Deploys K8S resources."""
+        try:
+            self.unit.status = MaintenanceStatus("Creating K8S resources")
+            self.k8s_resource_handler.apply()
+        except ApiError:
+            raise ErrorWithStatus("K8S resources creation failed", BlockedStatus)
+        self.model.unit.status = MaintenanceStatus("K8S resources created")
+
+    def _on_install(self, _):
+        """Perform installation only actions."""
```

Comment:
This felt a bit confusing.  I think I understand that you're trying to say "do the things for install that aren't container-specific".  

Maybe something like "Install supporting kubernetes resources." or similar?

Reply:
the ideal message is that you're deploying stuff that doesn't depend on the container being up.  but that's awkward to say

Reply:
I just added comment to the code and then I saw this. Here is the diff:
```
@@ -197,6 +197,7 @@ class JupyterUI(CharmBase):
     def _on_install(self, _):
         """Perform installation only actions."""
         try:
+            # deploy K8S resources to speed up deployment
             self._deploy_k8s_resources()
         except CheckFailed as err:
             self.model.unit.status = err.status

```

Reply:
Let me know if that's clear enough.

Reply:
I like the comment, but think it should be part of the docstring.  The docstring (in theory...) will show up in places where people don't have the rest of the code.  So just reading "Perform installation only actions." on it own doesn't feel like it tells me what's actually happening

Reply:
It will only display first line of that doc string. So, we won't be able to really explain all the steps taken in install. In this case it just creates K8S ahead of container being created. In other charms it could be something else.